bowjson.go

```diff
@@ -1,88 +1,173 @@
+
+	series := make([]Series, len(rec.Schema.Fields))
+	for i, f := range rec.Schema.Fields {
+		typ := newTypeFromArrowName(f.Type)
+		length := len(rec.Data)
+		buf, err := NewBufferFromInterfacesIter(
+			typ,
+			length,
+			func() chan interface{} {
 				cellsChan := make(chan interface{})
 				go func(cellsChan chan interface{}, colName string) {
-					for _, row := range jsonBow.Data {
+					for _, row := range rec.Data {
 						val, ok := row[colName]
 						if !ok {
 							cellsChan <- nil
 						} else {
 							_, ok = val.(float64)
-							if t == Int64 && ok {
+							if typ == Int64 && ok {
 								val = int64(val.(float64))
 							}
 							cellsChan <- val
 						}
 					}
 					close(cellsChan)
-				}(cellsChan, colSchema.Name)
+				}(cellsChan, f.Name)
+
 				return cellsChan
-			}())
-			if err != nil {
-				return err
-			}
-			series[i] = NewSeries(colSchema.Name, t, buf.Value, buf.Valid)
-			i++
-		}
-		tmpBow, err := NewBow(series...)
+			}(),
+		)
 		if err != nil {
-			return err
+			return nil, fmt.Errorf("%v: %v", ErrUnmarshalJSON, err)
+		}
+		series[i] = NewSeries(f.Name, typ, buf.Value, buf.Valid)
+	}
+
+	return NewBow(series...)
+}
+
+func EncodeBowToJSONBody(b Bow) (io.Reader, error) {
+	b.SetMarshalJSONRowBased(true)
+	jsonBody, err := b.MarshalJSON()
+	if err != nil {
+		return nil, fmt.Errorf("%v: %v", ErrEncodeBow, err)
+	}
+
+	return bytes.NewReader(jsonBody), nil
+}
+
+func DecodeJSONRespToBow(resp io.Reader) (Bow, error) {
+	respBytes, err := ioutil.ReadAll(resp)
+	if err != nil {
+		return nil, fmt.Errorf("%v: %v", ErrDecodeJSON, err)
+	}
+
+	var rec jsonRecord
+	err = json.Unmarshal(respBytes, &rec)
+	if err != nil {
+		return nil, fmt.Errorf("%v: %v", ErrDecodeJSON, err)
+	}
+
+	/*
+			Convert back json_table data types to bow data types
+
+			From pandas / io / json / _table_schema.py / as_json_table_type(x: DtypeObj) -> str:
+		    This table shows the relationship between NumPy / pandas dtypes,
+		    and Table Schema dtypes.
+		    ==============  =================
+		    Pandas type     Table Schema type
+		    ==============  =================
+		    int64           integer
+		    float64         number
+		    bool            boolean
+		    datetime64[ns]  datetime
+		    timedelta64[ns] duration
+		    object          str
+		    categorical     any
+		    =============== =================
+	*/
+
+	for i, f := range rec.Schema.Fields {
+		if _, ok := mapArrowDataTypeNameType[f.Type]; ok {
+			continue
 		}
-		b.Record = tmpBow.(*bow).Record
-		return nil
+		switch f.Type {
+		case "integer":
+			rec.Schema.Fields[i].Type = "int64"
+		case "number":
+			rec.Schema.Fields[i].Type = "float64"
+		case "boolean":
+			rec.Schema.Fields[i].Type = "bool"
+		}
+	}
+
+	jsonRec, err := json.Marshal(rec)
+	if err != nil {
+		return nil, fmt.Errorf("%v: %v", ErrDecodeJSON, err)
+	}
+
+	outputBow, err := UnmarshalJSON(jsonRec)
```

Comment:
marshalling then unmarshalling is unefficient. couln't you transform a pandaReadyJsonRecord to a Bow directly ?